controllers/securitypolicy_controller.go

```diff
@@ -50,135 +47,121 @@ type SecurityPolicyReconciler struct {
 // +kubebuilder:rbac:groups=tyk.tyk.io,resources=securitypolicies,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=tyk.tyk.io,resources=securitypolicies/status,verbs=get;update;patch
 
+// Reconcile reconciles SecurityPolicy custom resources
 func (r *SecurityPolicyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
 	log := r.Log.WithValues("SecurityPolicy", req.NamespacedName.String())
 
-	policyNamespacedName := req.NamespacedName.String()
-
+	ns := req.NamespacedName.String()
 	log.Info("Reconciling SecurityPolicy instance")
 
 	// Lookup policy object
-	desired := &tykv1.SecurityPolicy{}
-	if err := r.Get(ctx, req.NamespacedName, desired); err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(err) // Ignore not-found errors
+	policy := &tykv1.SecurityPolicy{}
+	if err := r.Get(ctx, req.NamespacedName, policy); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	// If object is being deleted
-	if !desired.ObjectMeta.DeletionTimestamp.IsZero() {
-		// The object is being deleted
-		if containsString(desired.ObjectMeta.Finalizers, securityPolicyFinalzerName) {
-			// our finalizer is present, so lets handle our external dependency
-			if err := r.UniversalClient.SecurityPolicy().Delete(desired.Status.PolID); err != nil {
-				log.Error(err, "unable to delete policy", "nameSpacedName", policyNamespacedName, "polId", desired.Status.PolID)
-				return reconcile.Result{}, nil
+	op, err := util.CreateOrUpdate(ctx, r.Client, policy, func() error {
+		if !policy.ObjectMeta.DeletionTimestamp.IsZero() {
+			if util.ContainsFinalizer(policy, policyFinalizer) {
+				return r.delete(ctx, policy)
 			}
-
-			err := r.UniversalClient.HotReload()
+			return nil
+		}
+		if !util.ContainsFinalizer(policy, policyFinalizer) {
+			util.AddFinalizer(policy, policyFinalizer)
+		}
```

Comment:
shouldn't these be outside the CreateOrUpdate function?

Reply:
I don't think so, we are updating the object, this will eventually need to trigger an update to reflect the changes.